src/cast_core/src/AssertiveActor.cpp

```diff
@@ -0,0 +1,219 @@
+        bsoncxx::document::view actualDoc{actualVal.get_document().value};
+        return equalBSONDocs(config, expectedDoc, actualDoc);
+    }
+    return actualVal.get_value() == expectedVal.get_value();
+}
+
+bool equalBSONDocs(AssertiveActor::PhaseConfig* config, const bsoncxx::document::view& expected, const bsoncxx::document::view& actual) {
+    for (auto actualVal : actual) {
+        auto key = actualVal.key();
+        if (config->ignoreFields.find(std::string(key)) != config->ignoreFields.end()) {
+            continue;
+        }
+        auto expectedVal = expected[key];
+        if (!equal<bsoncxx::document::element>(config, expectedVal, actualVal)) {
+            BOOST_LOG_TRIVIAL(info) << prefix(config) << "Documents differ in field: " << key;
+            return false;
+        }
+    }
+    for (auto expectedVal: expected) {
+        auto key = expectedVal.key();
+        if (config->ignoreFields.find(std::string(key)) != config->ignoreFields.end()) {
+            continue;
+        }
+        if (!actual[key]) {
+            BOOST_LOG_TRIVIAL(info) << prefix(config) << "Documents differ in field: " << key;
+            return false;
+        }
+    }
+    return true;
+}
+
+bool equalBSONArrays(AssertiveActor::PhaseConfig* config, const bsoncxx::array::view& expected, const bsoncxx::array::view& actual) {
+    auto expectedArrLen = std::distance(expected.begin(), expected.end());
+    auto actualArrLen = std::distance(actual.begin(), actual.end());
+    if (expectedArrLen != actualArrLen) {
+        BOOST_LOG_TRIVIAL(info) << prefix(config) << "Arrays differ in length. Expected " << expectedArrLen << " but found " << actualArrLen;
+        return false;
+    }
+
+    auto expectedIt = expected.begin();
+    int count = 0;
+    for (auto actualVal : actual) {
+        auto expectedVal = *expectedIt;
+        if (!equal<bsoncxx::array::element>(config, expectedVal, actualVal)) {
+            BOOST_LOG_TRIVIAL(info) << prefix(config) << "Arrays differ in idx " << count;
+            return false;
+        }
+        ++expectedIt;
+        ++count;
+    }
+
+    return true;
+}
+
+void AssertiveActor::run() {
+    for (auto&& config : _loop) {
+        for (const auto&& _ : config) {
+            auto expected = config->expectedExpr();
+            auto actual = config->actualExpr();
+
+            auto assertOp = _assert.start();
+            try {
+                auto expectedRes = runCommandAndGetResult(config, expected);
+                auto actualRes = runCommandAndGetResult(config, actual);
+                if (equalBSONArrays(config, getBatchFromCommandResult(expectedRes.view()), getBatchFromCommandResult(actualRes.view()))) {
+                    BOOST_LOG_TRIVIAL(info) << prefix(config) << "passed: results are equal.";
+                    assertOp.success();
+                } else {
+                    BOOST_THROW_EXCEPTION(FailedAssertionException("failed: results were unequal."));
```

Comment:
It might be a personal decision, but if there may be a dozen places where things are different, and only the first one is logged via exception throw, then it would take a dozen different long perf runs runs to track down all the differences. Seems better in such a case to print them all out.

That said, I could see someone wanting a fail-fast option. Deferring adding that until it's needed seems reasonable.